docs/get-details/dapps/smart-contracts/debugging.md

```diff
@@ -0,0 +1,223 @@
+
+The debugging session runs in a mock version of the ledger so the context including; balance records of the accounts, application parameters, and assets used during program evaluation must be supplied.  Exactly what context is needed for a debug session will depend on the type of contract that is being debugged and what opcodes are used in the program. 
+
+Most frequently, a `Dryrun Dump` file is used to pass all this context information in a single payload. 
+
+This file may be msgpack or json and can be created using goal or the SDKs
+
+
+=== "Python"
+    ```py
+
+    app_txn = transaction.ApplicationCallTxn(addr, sp, app_id, transaction.OnComplete.NoOpOC, accounts=[other_addr], foreign_assets=[asset_id])
+    s_app_txn = app_txn.sign(sk)
+
+    drr = transaction.create_dryrun(client, [s_app_txn])
+
+    filename = "dryrun.msgp"
+    with open(filename, "wb") as f:
+        import base64
+        f.write(base64.b64decode(encoding.msgpack_encode(drr)))
+
+    ```
+
+=== "JavaScript"
+    ```js
+
+    const app_txn = algosdk.makeApplicationNoOpTxn(addr, sp, app_id, undefined, [other_addr], undefined, undefined)
+    const s_app_txn = algosdk.signTransaction(app_txn, sk)
+
+    const drr = await algosdk.createDryrun({
+        client: client, 
+        txns: [
+            algosdk.decodeSignedTransaction(s_app_txn['blob']),
+        ]
+    })
+
+    const filename = 'dryrun.msgp'
+    fs.writeFileSync(filename, algosdk.encodeObj(drr.get_obj_for_encoding(true)))
+
+    ```
+
+=== "Go"
+    ```go
+
+	app_txn, err := future.MakeApplicationNoOpTx(app_id, nil, []string{other_addr}, nil, []uint64{asset_id}, sp, addr, nil, types.Digest{}, [32]byte{}, types.Address{})
+	if err != nil {
+		log.Fatalf("Failed to create app call txn: %+v", err)
+	}
+
+    _, s_app_bytes, err := crypto.SignTransaction(sk, app_txn)
+	if err != nil {
+		log.Fatalf("Failed to sign app txn: %+v", err)
+	}
+	s_app_txn := types.SignedTxn{}
+	msgpack.Decode(s_app_bytes, &s_app)
+
+    drr, err := future.CreateDryrun(client, []types.SignedTxn{s_app_txn}, nil, context.Background())
+	if err != nil {
+		log.Fatalf("Failed to create dryrun: %+v", err)
+	}
+
+	filename := "dryrun.msgp"
+	os.WriteFile(filename, msgpack.Encode(drr), 0666)
+
+    ```
+
+=== "Java"
+    ```java
+
+    Transaction app_txn = ApplicationCallTransactionBuilder.Builder().sender(addr)
+        .suggestedParams(sp).applicationId(app_id).foreignAssets(fassets).accounts(addrs).build();
+
+
+    List<SignedTransaction> stxns = new ArrayList<>();
+    stxns.add(acct.signTransaction(app_txn));
+
+    DryrunRequest drr = Utils.createDryrun(client, stxns, "", 0L, 0L);
+
+    String fname = "dryrun.msgp";
+    FileOutputStream outfile = new FileOutputStream(fname);
+    outfile.write(Encoder.encodeToMsgPack(drr));
+    outfile.close();
+
+    ```
+
+=== "goal"
+    ```sh
+
+    $ goal app call --app-id {appid} --from {ACCOUNT} --out=dryrun.json --dryrun-dump
```

Comment:
```suggestion
    $ goal app call --app-id {appid} --from {ACCOUNT} --out=dryrun.json --dryrun-dump
    # or if you already have a transaction file
    $ goal clerk dryrun --dryrun-dump -t transaction.txn -o dryrun.json
```

Reply:
I realise you have the alternative way a little further down, so no problem ignoring this.